libcaf_net/caf/net/packet_writer.hpp

```diff
@@ -0,0 +1,60 @@
+/******************************************************************************
+ *                       ____    _    _____                                   *
+ *                      / ___|  / \  |  ___|    C++                           *
+ *                     | |     / _ \ | |_       Actor                         *
+ *                     | |___ / ___ \|  _|      Framework                     *
+ *                      \____/_/   \_|_|                                      *
+ *                                                                            *
+ * Copyright 2011-2019 Dominik Charousset                                     *
+ *                                                                            *
+ * Distributed under the terms and conditions of the BSD 3-Clause License or  *
+ * (at your option) under the terms and conditions of the Boost Software      *
+ * License 1.0. See accompanying files LICENSE and LICENSE_ALTERNATIVE.       *
+ *                                                                            *
+ * If you did not receive a copy of the license files, see                    *
+ * http://opensource.org/licenses/BSD-3-Clause and                            *
+ * http://www.boost.org/LICENSE_1_0.txt.                                      *
+ ******************************************************************************/
+
+#pragma once
+
+#include <vector>
+
+#include "caf/byte.hpp"
+#include "caf/net/fwd.hpp"
+#include "caf/span.hpp"
+
+namespace caf {
+namespace net {
+
+/// Implements an interface for packet writing in application-layers.
+class packet_writer {
+public:
+  using buffer_type = std::vector<byte>;
+
+  virtual ~packet_writer() = default;
+
+  /// Returns the next header_buffer from transport
+  virtual buffer_type next_header_buffer() = 0;
+
+  /// Returns the next payload_buffer from transport
+  virtual buffer_type next_buffer() = 0;
```

Comment:
I think we should make the API a bit clearer:

```suggestion
  virtual buffer_type next_payload_buffer() = 0;
```